src/pybind/mgr/restful/decorators.py

```diff
@@ -0,0 +1,56 @@
+from pecan import request, response
+from base64 import b64decode
+from functools import wraps
+
+import traceback
+
+import module
+
+
+# Handle authorization
+def auth(f):
+    @wraps(f)
+    def decorated(*args, **kwargs):
+        if not request.authorization:
+            response.status = 401
+            response.headers['WWW-Authenticate'] = 'Basic realm="Login Required"'
+            return {'message': 'auth: No HTTP username/password'}
+
+        username, password = b64decode(request.authorization[1]).split(':')
+
+        # Check that the username exists
+        if username not in module.instance.keys:
+            response.status = 401
+            response.headers['WWW-Authenticate'] = 'Basic realm="Login Required"'
+            return {'message': 'auth: No such user'}
+
+        # Check the password
+        if module.instance.keys[username] != password:
+            response.status = 401
+            response.headers['WWW-Authenticate'] = 'Basic realm="Login Required"'
+            return {'message': 'auth: Incorrect password'}
+
+        return f(*args, **kwargs)
+    return decorated
+
+
+# Helper function to catch and log the exceptions
+def catch(f):
```

Comment:
I switched to the on_error hook with the latest commit. The catch function had the benefit of returning the traceback directly to the user, not just to the ceph-mgr log but that was mostly a convenience hack for me while I was testing this.